samza-core/src/main/java/org/apache/samza/job/metadata/JobCoordinatorMetadataManager.java

```diff
@@ -207,29 +213,33 @@ public void writeJobCoordinatorMetadata(JobCoordinatorMetadata metadata) {
   }
 
   /**
-   * Generate the epoch id using the execution container id that is passed through system environment. This isn't ideal
-   * way of generating this ID and we will need some contract between the underlying cluster manager and samza engine
-   * around what the epoch ID should be like and what is needed to generate is across different cluster offerings.
-   * Due to unknowns defined above, we leave it as is and keep it simple for now. It is favorable to keep it this way
-   * instead of introducing a loosely defined interface/API and marking it unstable.
-   *
    * The properties of the epoch identifier are as follows
    *  1. Unique across applications in the cluster
    *  2. Remains unchanged within a single deployment lifecycle
    *  3. Remains unchanged across application attempt within a single deployment lifecycle
    *  4. Changes across deployment lifecycle
    *
-   *  Note: The above properties is something we want keep intact when extracting this into a well defined interface
-   *  or contract for YARN AM HA to work.
-   *  The format and property used to generate ID is specific to YARN and the specific format of the container name
-   *  is a public contract by YARN which is likely to remain backward compatible.
+   * For YARN environment:
+   * Generate the epoch id using the execution container id that is passed through system environment. This isn't ideal
+   * way of generating this ID, since it is YARN-specific. This is left as a legacy flow for backwards compatibility, as
+   * the original implementation did not define a cluster-agnostic contract.
+   * The format and property used to generate ID is specific to YARN and the specific format of the container name
+   * is a public contract by YARN which is likely to remain backward compatible.
+   *
+   * For non-YARN environments:
+   * Extract this from the environment variable SAMZA_EPOCH_ID. This is a more generic way of obtaining an epoch id, but
+   * it does require the resource management layer to inject this value.
    *
    * @return an identifier associated with the job coordinator satisfying the above properties
    */
   @VisibleForTesting
   String fetchEpochIdForJobCoordinator() {
-    String[] containerIdParts = getEnvProperty(CONTAINER_ID_PROPERTY).split(CONTAINER_ID_DELIMITER);
-    return containerIdParts[1] + CONTAINER_ID_DELIMITER + containerIdParts[2];
+    if (ClusterType.YARN.equals(this.clusterType)) {
+      String[] containerIdParts = getEnvProperty(CONTAINER_ID_PROPERTY).split(CONTAINER_ID_DELIMITER);
+      return containerIdParts[1] + CONTAINER_ID_DELIMITER + containerIdParts[2];
+    } else {
+      return getEnvProperty("SAMZA_EPOCH_ID");
```

Comment:
Is it fair to treat as this a configuration that operators of samza if not the end users should be aware of? If so, can we document what is the expected properties, format/protocol of the value? 

Unfortunately, we don't have a central place for some of the environment variables we pass to our system and i am okay if you want to document it alongside the configurations or if you find some other place that is appropriate.

Reply:
Good point, I'll improve docs on this.